klippy/webhooks.py

```diff
@@ -0,0 +1,245 @@
+
+    def put(self, name, value):
+        self.args[name] = value
+
+    def get_int(self, item):
+        return int(self.get(item))
+
+    def get_float(self, item):
+        return float(self.get(item))
+
+    def get_args(self):
+        return self.args
+
+    def get_path(self):
+        return self.path
+
+    def get_method(self):
+        return self.method
+
+    def set_error(self, error):
+        self.response = error.to_dict()
+
+    def send(self, data):
+        if self.response is not None:
+            raise WebRequestError("Multiple calls to send not allowed")
+        self.response = data
+
+    def finish(self):
+        if self.response is None:
+            # No error was set and the user never executed
+            # send, default response is "ok"
+            self.response = "ok"
+        return {"request_id": self.id, "response": self.response}
+
+class ServerConnection:
+    def __init__(self, webhooks, printer):
+        self.webhooks = webhooks
+        self.reactor = printer.get_reactor()
+
+        # Klippy Connection
+        self.fd_handler = self.mutex = None
+        self.is_server_connected = False
+        self.partial_data = ""
+        is_fileoutput = (printer.get_start_args().get('debugoutput')
+                         is not None)
+        if is_fileoutput:
+            # Do not try to connect in klippy batch mode
+            return
+        self.socket = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
+        self.socket.setblocking(0)
+        try:
+            self.socket.connect(SOCKET_LOCATION)
+        except socket.error:
+            logging.debug(
+                "ServerConnection: Moonraker server not detected")
+            return
+        logging.debug("ServerConnection: Moonraker connection established")
+        self.is_server_connected = True
+        self.fd_handler = self.reactor.register_fd(
+            self.socket.fileno(), self.process_received)
+        self.mutex = self.reactor.mutex()
+        printer.register_event_handler('klippy:disconnect', self.close_socket)
+
+    def close_socket(self):
+        if self.is_server_connected:
+            logging.info("ServerConnection: lost connection to Moonraker")
+            self.is_server_connected = False
+            self.reactor.unregister_fd(self.fd_handler)
+            try:
+                self.socket.close()
+            except socket.error:
+                pass
+
+    def is_connected(self):
+        return self.is_server_connected
+
+    def process_received(self, eventtime):
+        try:
+            data = self.socket.recv(4096)
+        except socket.error as e:
+            # If bad file descriptor allow connection to be
+            # closed by the data check
+            if e.errno == 9:
+                data = ''
+            else:
+                return
+        if data == '':
+            # Socket Closed
+            self.close_socket()
+            return
+        requests = data.split('\x00')
```

Comment:
What do you think about separating by '\x03' (the old ASCII ETX / end of message indictor) instead of by null?  A null character can be a bit of a hassle to work with in some languages.